### PR TITLE
tests(opinions): Add Testing method for extract from text

### PR DESCRIPTION
**Pull request by flooie**
Add testing for extract_from_text

The test requires a test file for any module with extract from text implemented.

Its similar to our regular testing infrastructure.  In this case though its one test 
input for each entry in the compare json file.

### CODE REVIEW

File: tests/local/test_ScraperExtractFromTextTest.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+class ScraperExtractFromText(unittest.TestCase):
+    """Adds specific tests to specific courts that are more-easily tested
+    without a full integration test.
+    """
+
+    def run_tests_on_module_str(self, module_str: str) -> List[str]:
+        """Finds all the $module_example* files and tests them with the sample
+        scraper.
+        """
+        return build_module_list(module_str)
```
Comment (mlissner) [inline tests/local/test_ScraperExtractFromTextTest.py:19]: If this is just a wrapper for a one-liner, why bother with it at all?

File: tests/local/test_ScraperExtractFromTextTest.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                pass
+
+            module_parts = module_string.split(".")
+            example_path = os.path.join(
+                "tests",
+                "examples",
+                module_parts[1],
+                "united_states",
+                module_parts[-1],
+            )
```
Comment (mlissner) [inline tests/local/test_ScraperExtractFromTextTest.py:47]: You should start leaning towards using `Pathlib` for this kind of stuff. The above would be something like:

```suggestion
            example_path = Path("tests") / "examples" / module_parts[1] / "united_states" / module_parts[-1]
```

(Of course, Black will have something to say about this, and I admit it's not the biggest improvement, so no biggie if you want to just ignore this, but at this point I don't think there are many [any?] good reasons for using `os.path.join`)

File: tests/local/test_ScraperExtractFromTextTest.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            # Run the compare file for any module (non-backscraper) that has extract_from_text implemented.
+            with open(f"{example_path}_extract_from_text.compare.json") as f:
+                answers = json.load(f)
+
+            sorted_paths = sorted(
+                glob(f"{example_path}_extract_from_text*.txt")
```
Comment (mlissner) [inline tests/local/test_ScraperExtractFromTextTest.py:53]: Probably should throw this in a variable so you don't have to run the same code a second time a few lines down.

File: tests/local/test_ScraperExtractFromTextTest.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            )
+
+            self.assertEqual(
+                len(sorted_paths),
+                len(answers),
+                msg=f"Missing test files for extract_from_text in {module_parts[-1]}",
+            )
+
+            for file in sorted(glob(f"{example_path}_extract_from_text*.txt")):
+                s = site.extract_from_text(open(file).read())
```
Comment (mlissner) [inline tests/local/test_ScraperExtractFromTextTest.py:63]: Recommend context handler here (`with open(file) as f...`)

File: tests/local/test_ScraperExtractFromTextTest.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+            sorted_paths = sorted(
+                glob(f"{example_path}_extract_from_text*.txt")
+            )
+
+            self.assertEqual(
+                len(sorted_paths),
+                len(answers),
+                msg=f"Missing test files for extract_from_text in {module_parts[-1]}",
+            )
```
Comment (mlissner) [inline tests/local/test_ScraperExtractFromTextTest.py:60]: it's a little unusual to have one answers file and multiple test files. There a reason to do this differently than elsewhere?

File: tests/local/test_ScraperExtractFromTextTest.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        for module_string in module_strings:
+            if "backscraper" in module_string:
+                continue
+            package, module = module_string.rsplit(".", 1)
+            mod = __import__(
+                f"{package}.{module}", globals(), locals(), [module]
+            )
+            site = mod.Site()
+            try:
+                if site.extract_from_text("This is a test.") == {}:
```
Comment (mlissner) [inline tests/local/test_ScraperExtractFromTextTest.py:34]: Unless I'm mistaken, wouldn't the `extract_from_text` method return an empty dict if, say, it didn't find the text it was looking for — for example, if the fixture was no good? 

Assuming that's the case, would it make sense to actually check if the method is overridden (instead of based on its outputs)? I [think](https://stackoverflow.com/a/44914111/64911) you can do so with something like:

```
if Site.extract_from_text == OpinionSite.extract_from-text:
    # not overridden
    continue
```

File: tests/examples/opinions/united_states/bia_extract_from_text_2.txt [hunk 7]
```diff
@@ -0,0 +1,41 @@
+Cite as 27 I&N Dec. 48 (BIA 2017) Interim Decision #3894
```
Comment (mlissner) [inline tests/examples/opinions/united_states/bia_extract_from_text_2.txt:1]: These fixtures are big. Can you winnow them down to just the relevant parts?

File: juriscraper/opinions/united_states/administrative_agency/bia.py [hunk 8]
```diff
@@ -118,7 +118,7 @@ def extract_from_text(self, scraped_text: str) -> Dict[str, Any]:
         date = re.findall(
             r"Decided (by (Acting\s)?Attorney General )?(.*\d{4})",
             scraped_text,
-        )[0][1]
+        )[0][-1]
```
Comment (mlissner) [inline juriscraper/opinions/united_states/administrative_agency/bia.py:121]: Um? This is a bit sus, no?

**Review by mlissner — CHANGES_REQUESTED** [review]
More comments than you probably expected, but the design is sound and looks about like what I'd expect. Just some quibbles about the details, mostly.

File: tests/examples/opinions/united_states/bia_extract_from_text_2.txt (hunk 7 above)
Comment (flooie) [reply tests/examples/opinions/united_states/bia_extract_from_text_2.txt:1]: yes

File: tests/local/test_ScraperExtractFromTextTest.py (hunk 6 above)
Comment (mlissner) [reply tests/local/test_ScraperExtractFromTextTest.py:34]: Oh, and sorry, one other thing, if it's the case that it *is* overridden, I don't think I see how we enforce that there are tests. Could we also assert that the number of test cases is greater than zero if the method is overridden?

File: tests/examples/opinions/united_states/bia_extract_from_text_2.txt (hunk 7 above)
Comment (flooie) [reply tests/examples/opinions/united_states/bia_extract_from_text_2.txt:1]: Im always somewhat concerned that a fixture or regex pattern might be thrown off by a large PDF document, so I always make them larger as proof but I understand that they are large.  If we are shrinking them then I think it might be smarter to just figure a different way to do this.

File: juriscraper/opinions/united_states/administrative_agency/bia.py (hunk 8 above)
Comment (flooie) [reply juriscraper/opinions/united_states/administrative_agency/bia.py:121]: suspect?  ... 

Making the test let me find a bug in my regex pattern.  I was trying to sneak in the fix for BIA

File: tests/local/test_ScraperExtractFromTextTest.py (hunk 6 above)
Comment (flooie) [reply tests/local/test_ScraperExtractFromTextTest.py:34]: No.  It wouldn't return {}.  

In the BIA context if it didn't it would return. Either NONE because it failed or return the following dict where date_filed would be incorrect I think.  Either way an empty {} is not what would be returned.  

```
metadata = {
            "OpinionCluster": {
                "date_filed": date_filed,
                "date_filed_is_approximate": False,
            },
        }
```

File: tests/local/test_ScraperExtractFromTextTest.py (hunk 1 above)
Comment (flooie) [reply tests/local/test_ScraperExtractFromTextTest.py:19]: Mostly modeling from other tests

Comment (mlissner) [reply tests/local/test_ScraperExtractFromTextTest.py:19]: Uh oh, that probably means past-Mike wrote those tests, huh? He's the worst.

File: juriscraper/opinions/united_states/administrative_agency/bia.py (hunk 8 above)
Comment (mlissner) [reply juriscraper/opinions/united_states/administrative_agency/bia.py:121]: Sus == suspect. Gotta hang with the kids, Bill!

Fix makes sense. Glad to see the new test working!

File: tests/examples/opinions/united_states/bia_extract_from_text_2.txt (hunk 7 above)
Comment (mlissner) [reply tests/examples/opinions/united_states/bia_extract_from_text_2.txt:1]: Not a strong feeling here. I appreciate small ones because they focus my attention on what they're trying to test, rather than the gazillion other things (and they keep repos small, etc). But really, just a slight preference here. Maybe I shouldn't care.

File: tests/local/test_ScraperExtractFromTextTest.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                    continue
+            except:
```
Comment (mlissner) [inline tests/local/test_ScraperExtractFromTextTest.py:36]: This shouldn't be so broad, BTW. Just noticed.

File: tests/local/test_ScraperExtractFromTextTest.py (hunk 6 above)
Comment (mlissner) [reply tests/local/test_ScraperExtractFromTextTest.py:34]: OK, so I'm confused what this try/if/except block is doing, in that case, sorry?

File: tests/local/test_ScraperExtractFromTextTest.py (hunk 9 above)
Comment (flooie) [reply tests/local/test_ScraperExtractFromTextTest.py:36]: broad?

File: tests/local/test_ScraperExtractFromTextTest.py (hunk 6 above)
Comment (flooie) [reply tests/local/test_ScraperExtractFromTextTest.py:34]: If I pass in a random string.  if we haven't implemented it -... its going to return `{}`. so skip.  But if we have implemented it ... its going to crash.

File: tests/local/test_ScraperExtractFromTextTest.py (hunk 1 above)
Comment (flooie) [reply tests/local/test_ScraperExtractFromTextTest.py:19]: fair enough... ill drop this

File: tests/examples/opinions/united_states/bia_extract_from_text_2.txt (hunk 7 above)
Comment (flooie) [reply tests/examples/opinions/united_states/bia_extract_from_text_2.txt]: √

**Comment by flooie** [comment]
@mlissner I changed the tests completely.  I dropped the testing files, I think it doesn't make sense (in hindsight) to do it in the same way as the rest of the opinion tests.  It muddies it up too much.  Instead I have two tests. One that validates that every implemented extract from text is in fact tested.  And the other to do the tests.

File: juriscraper/opinions/united_states/administrative_agency/bia.py (hunk 8 above)
Comment (flooie) [reply juriscraper/opinions/united_states/administrative_agency/bia.py:121]: oh dont worry my wife brings home the lingo from the kids

**Comment by mlissner** [comment]
Thank you.
